### ISSUE TITLE
Will the observer(for activation) be synchronized across GPUs in DDP?

### ISSUE THREAD

**Issue by WZMIAOMIAO** [pytorch/pytorch#85627]
### 🚀 The feature, motivation and pitch

First of all, thanks for your great work. I have learned about the quantization code of this repo these days.  And I have a question that In the observer implementation, I don't see any code related to synchronization of multiple GPU. My personal opinion is that if we use DDP to quantize model, we need to synchronize the observer's `min_val` and `max_val` for activation, because different processes read different batch data. Since the model weights of different GPUs are the same, the observer‘s `min_val` and `max_val` for weights  does not need to be synchronized. If my understanding is wrong, please point out. Thanks.

### Demo

`main.py`
```python
import os
from argparse import Namespace
import torch
import torch.distributed as dist


def init_distributed_mode(args):
    if 'RANK' in os.environ and 'WORLD_SIZE' in os.environ:
        args.rank = int(os.environ["RANK"])
        args.world_size = int(os.environ['WORLD_SIZE'])
        args.gpu = int(os.environ['LOCAL_RANK'])
    elif 'SLURM_PROCID' in os.environ:
        args.rank = int(os.environ['SLURM_PROCID'])
        args.gpu = args.rank % torch.cuda.device_count()
    elif hasattr(args, "rank"):
        pass
    else:
        print('Not using distributed mode')
        args.distributed = False
        return

    args.distributed = True

    torch.cuda.set_device(args.gpu)
    args.dist_backend = 'nccl'
    print('| distributed init (rank {}): {}'.format(
        args.rank, args.dist_url), flush=True)
    torch.distributed.init_process_group(backend=args.dist_backend, init_method=args.dist_url,
                                         world_size=args.world_size, rank=args.rank)


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1, 1)
        self.register_buffer("max_val", torch.tensor(float("-inf")))

    def forward(self, x):
        self.max_val = torch.amax(x)
        x = self.conv(x)
        return x


def main():
    args = Namespace(dist_url="env://")
    init_distributed_mode(args)

    if args.distributed is False:
        return

    model = M()
    model.cuda()
    model = torch.nn.parallel.DistributedDataParallel(model)

    x = torch.full(size=(1, 1, 2, 2), fill_value=float(args.gpu)).cuda()
    model(x)
    
    print(f"idx: {args.gpu}, max_val: {model.module.max_val.item()}")


if __name__ == '__main__':
    main()

```

run command:
```
torchrun --nproc_per_node=2 main.py
```
outputs:
```
idx: 0, max_val: 0.
idx: 1, max_val: 1.
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @pietern @SciPioneer

**Comment by awgu** [comment]
Hi! I agree with your assessment. The `min_val` and `max_val` for model parameters (i.e. weights) _do not_ need to be synchronized because the the parameters will be the same across ranks (due to the gradient synchronization). In contrast, the `min_val` and `max_val` for activations _do_ need to be synchronized because the ranks see different input data.

In your case, you may need to call `dist.all_reduce(self.max_val, op=ReduceOp.MAX)` on all ranks somewhere in your training code each iteration.

**Comment by awgu** [comment]
I have an example of how you might do this:
```
import os
from typing import Dict

import torch
import torch.distributed as dist
import torch.multiprocessing as mp
from torch._C._distributed_c10d import ReduceOp
from torch.nn.parallel import DistributedDataParallel as DDP


class M(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.conv = torch.nn.Conv2d(1, 1, 1, 1)
        self.register_buffer("max_val", torch.tensor(float("-inf")))

    def forward(self, x):
        self.max_val = torch.amax(x)
        x = self.conv(x)
        return x

    @staticmethod
    def all_reduce_max_val_hook(ddp: DDP, named_buffers: Dict[str, torch.Tensor]):
        assert "max_val" in named_buffers
        dist.all_reduce(named_buffers["max_val"], op=ReduceOp.MAX)


def main(rank: int, world_size: int):
    dist.init_process_group("nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

    model = M()
    model.cuda()
    model = DDP(model, device_ids=[rank])
    hook_post_fwd = torch.nn.parallel.distributed._BufferCommHookLocation.POST_FORWARD
    model._register_buffer_comm_hook(
        model, M.all_reduce_max_val_hook, hook_post_fwd,
    )

    x = torch.full(size=(1, 1, 2, 2), fill_value=float(rank)).cuda()
    model(x)
    
    print(f"idx: {rank}, max_val: {model.module.max_val.item()}")


if __name__ == '__main__':
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "29500"
    world_size = 2
    mp.spawn(
        main,
        args=(world_size,),
        nprocs=world_size,
        join=True,
    )
```
I adapted the script to use `torch.multiprocessing.spawn()` for my own convenience, but it should not matter. The relevant code is the `_register_buffer_comm_hook()` call and the `all_reduce_max_val_hook()` definition. In the above example, each rank's `M.max_val` will be synchronized after the forward pass.

If `M` also had a `min_val` attribute, then you can add a line like `dist.all_reduce(named_buffers["min_val"], op=ReduceOp.MIN)` in `all_reduce_max_val_hook()`.

**Comment by WZMIAOMIAO** [comment]
thank you for your reply, I hope the official code can be supported natively.

**Comment by awgu** [comment]
I am marking this as closed then. Feel free to reopen if there is still any further question.
